sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -84,26 +91,102 @@ static RT::PiProgram createSpirvProgram(const RT::PiContext Context,
   return Program;
 }
 
+template<typename RetT>
+RetT *
+waitUntilBuilt(kernel_program_cache &Cache,
+               kernel_program_cache::EntityWithBuildState<RetT> *WithBuildState) {
+  // condition_variable.wait() call is wrapped with a while due to common
+  // cond_var for building all kernels and programs
+  int state = WithBuildState->State.load();
+  while (state != BS_Built && state != BS_BuildFailed) {
+    Cache.waitUntilBuilt([WithBuildState]() {
+      auto state = WithBuildState->State.load();
+
+      return state == BS_Built || state == BS_BuildFailed;
+    });
+
+    state = WithBuildState->State.load();
+  }
+
+  // expected to be null if state eq BuildFailed
+  return WithBuildState->Ptr.load();
+}
+
 RT::PiProgram ProgramManager::getBuiltOpenCLProgram(OSModuleHandle M,
                                                     const context &Context) {
   std::shared_ptr<context_impl> Ctx = getSyclObjImpl(Context);
-  std::map<OSModuleHandle, RT::PiProgram> &CachedPrograms =
-      Ctx->getCachedPrograms();
-  auto It = CachedPrograms.find(M);
-  if (It != CachedPrograms.end())
-    return It->second;
 
+  using PiProgramT = kernel_program_cache::PiProgramT;
+  using ProgramWithBuildState = kernel_program_cache::ProgramWithBuildStateT;
+
+  kernel_program_cache &Cache = *Ctx->getKernelProgramCache().get();
+
+  ProgramWithBuildState *ProgramWithState;
+  bool InsertionTookPlace;
```

Comment:
```suggestion
  bool InsertionTookPlace = false;
```

Reply:
It's initialized pretty soon after allocation. Don't see any if this change is obligatory